docs/requirements/localhost-requirements.md

```diff
@@ -0,0 +1,70 @@
+<!-- More detailed information about the requirements engineering process can be found at https://github.com/cosmos/ibc-go/wiki/Requirements-engineering -->
+
+# Business requirements
+
+The local host client provides a unified interface to interact with different applications on a single chain.
+
+## Problem
+
+Currently applications, or smart contracts, on a single chain cannot communicate through a single interface. This means a user must interact with each application or smart contract directly which can be cumbersome, especially when a user wants to interact with many applications on one chain.
+
+## Objectives
+
+To provide a single interface in line with IBC application layer semantics, that enables a user to interact with multiple different applications or smart contracts remotely on a given blockchain. 
+
+## Scope
+
+| Features  | Release |
+| --------- | ------- |
+| To interface with applications or smart contracts on a blockchain through the IBC application layer | v1 |
+| To use the localhost client without requiring third party relayer infrastructure | v2 |
+
+# User requirements
+
+## Use cases
+
+### Interacting with Smart Contracts
+
+Many IBC connected blockchains utilise smart contracts, these could be written in rust, solidity or javascript depending on the smart contract platform being used.  Through the localhost client, a user can interact with and call into smart contracts by sending ibc messages from the localhost client to the smart contract. 
+
+On Agoric, a planned use case is for managing delegations and staking with smart contracts on a given chain. MsgDelegate, MsgUndelegate, MsgBeginRedelegate messages would be sent using the localhost client to the staking module. The same interface could be used to manage staking cross chain, sending the same messages through IBC to an interchain account.
+
+### Interoperability Infrastructure
+
+Polymer plans to leverage the localhost client with multiple connections as part of their architecture to connect chains not using Tendermint or the Cosmos SDK to IBC enabled chains. Polymer will act as a hub connecting multiple blockchains together. 
+
+# Functional requirements
+
+## Assumptions and dependencies
+
+1. Chains utilising the localhost client must be using an ibc-go release in the v7 line.
+2. The channel behaviour of a localhost client will be as ICS 04 specifies.
+
+## Features
+
+### 1 - Configuration
+
+| ID | Description | Verification | Status | 
+| -- | ----------- | ------------ | ------ | 
+| 1.01 | The localhost client shall have a client ID of the string `09-localhost`. | ------------ | Drafted | 
+| 1.02 | The localhost client shall have a sentinel connection ID of the string `connection-localhost`. | ------------ | Drafted | 
+| 1.03 | If more than 1 localhost connection is required, this is possible with a different connection ID. | ------------ | Drafted |
```

Comment:
I think there is more work required to support this.

- Remove `len(proof) != 0` checks from connection handshake msgs `ValidateBasic` 
- Determine how connection handshake will handle [`VerifyClientState` and `VerifyClientConsensusState`](https://github.com/cosmos/ibc-go/blob/main/modules/core/03-connection/keeper/handshake.go#L122-L131) for localhost.

Note point 2 may nullify 1.04 here, as if we are supporting connection handshakes we must verify `expectedConsensusState` in `VerifyClientConsensusState` linked above. This calls into `VerifyMembership` and would fail trying to verify a `nil` value.

Reply:
cc. @AdityaSripal

Reply:
The first point is true, this PR would basically be extended to include the 03-connection messages as well

The verify point is a good question:

The verify messages could do the same as what they already do which is to delegate to the client layer.
The relayer should pass in the localhost client state and nil for the consensus state.
I think this would work if the tx is included at the same block the relayer passes in, which would be quite brittle.

Perhaps we need some specialcase handling here?

Reply:
I updated the requirement in 1.03 to require only 1 connection following consultation with Polymer that they are happy to handle this special case